draft-ZIP-0226.rst

```diff
@@ -0,0 +1,303 @@
+
+
+:math:`\mathsf{assetBurn = [(v^{type}, type_\mathbb{P})}| \forall \mathsf{type}_\mathbb{P}  \textit{ s.t.}\mathsf{v^{type}\neq 0}]`
+
+The value balances for each asset type in `assetBurn` represents the amount of that asset type that is being burnt. In the case of ZEC, the value balance represents either the transaction fee, or the amount of ZEC changing anonymity pools (to Sapling or Transparent).
+
+Finally, the validator needs to verify the Balance and Binding Signature by adding the value balances for all assets, as committed using their respective types as the value base point of the Pedersen Commitment. This is done as follows
+
+:math:`\mathsf{bvk = (\sum cv_i^{net})}  - \mathsf{ ValueCommit_0^{Orchard}(v^{balanceOrchard})} - \sum_{\forall \mathsf{type}\textit{ s.t. }\mathsf{v^{type}\neq 0}} \mathsf{Value Commit_0^{ZSA}(v^{type}type_\mathbb{P}) } = \sum \mathsf{rcv_{i,j}^{net}}\mathcal{R}^{\mathsf{Orchard}}`
+
+In the case that the balance of all the action values related to a specific asset will be zero, there will be no value added to the vector. This way, the number of assets, nor their types will be revealed, except in the case that an asset is burnt.
+
+**Note:** Even if this mechanism allows having transparent ↔  shielded asset transfers in theory, the transparent protocol will not be changed with this ZIP to adapt to a multiple asset structure. This means that unless future consensus rules changes do allow it, the unshielding is not not be possible for custom assets.
+
+ZSA Transaction Structure
+=========================
+Similar to NU5 transaction structure, with the following modifications to the Orchard bundle, as defined in [#protocol-transactionstructure]_:
+
++-----------------+-------------+-----------------------------------+-------------------------+
+| Bytes           | Name        | Data Type                         | Description             |
++=================+=============+===================================+=========================+
+| newActionSize * | vActionsZSA | ActionDescription[nActionOrchard] |                         |
+| nActionsZSA     |             |                                   |                         |
++-----------------+-------------+-----------------------------------+-------------------------+
+| varies          | nAssetBurn  | compactSize                       | number of assets burnt  |
++-----------------+-------------+-----------------------------------+-------------------------+
+| 40*nAssetBurn   | vAssetBurn  | bytes[40][nAssetBurn]             | 32 bytes asset type_t,  |
+|                 |             |                                   | 8 bytes of valueBalance |
```

Comment:
If we want to support wrapped ETH, we might want a 128-bit value balance here, to support wei (10^18^ ~= 2^68^).

But if we want to support any ERC20 asset, we might want 256-bit value balance.

See more details here from a developer who has tried to adapt Sapling to wrap ETH:
https://discord.com/channels/809218587167293450/1007021690510921728/1012505609926611114

Reply:
A 256-bit value balance would be tricky to support because it would invalidate the argument that value balances do not overflow. We should look more carefully at which ERC20 assets this would exclude.

Reply:
The note commitment input includes a 64-bit encoding of the note value. We could include the high 64 bits in the same extension that is used for the type of non-ZEC notes.